### DISCUSSION TITLE
Help: Getting "JsonSerializationException: Unable to find a constructor to use" error when trying to deserialize on WebGL

### DISCUSSION THREAD

**Discussion by JesseDunlap** [applejag/Newtonsoft.Json-for-Unity#68] [Q&A] [answered]
## Expected behavior

I have a basic JSON payload that I am attempting to deserialize into a known class.

I would expect to be able to convert the following JSON payload:
```
{"token":"f61886be-6d2b-49ec-81da-1cb184729849","expiresAt":"2020-10-29T17:58:36.408293Z","createdAt":"2020-10-14T17:58:36.408293Z"}
```

Into an instance of the following class...
```
public class AccessTokenDto
{
    public string Token { get; set; }

    public DateTime ExpiresAt { get; set; }
    
    public DateTime CreatedAt { get; set; }
}
```

By using the following code:
```
var accessTokenDto = JsonConvert.DeserializeObject<AccessTokenDto>(jsonText);
```

## Actual behavior

This works fine when running in the Unity editor, but when I build for WebGL, I receive the following exception in the console...

```
e23d7cc7-c1d6-4af3-9a6a-c148fc05331b:3472 JsonSerializationException: Unable to find a constructor to use for type Nebula.Grid.Common.Data.AccessTokenDto. A class should either have a default constructor, one constructor with arguments or a constructor marked with the JsonConstructor attribute. Path 'token', line 1, position 9.
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateNewObject (Newtonsoft.Json.JsonReader reader, Newtonsoft.Json.Serialization.JsonObjectContract objectContract, Newtonsoft.Json.Serialization.JsonProperty containerMember, Newtonsoft.Json.Serialization.JsonProperty containerProperty, System.String id, System.Boolean& createdFromNonDefaultCreator) [0x00000] in <00000000000000000000000000000000>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x00000] in <00000000000000000000000000000000>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x00000] in <00000000000000000000000000000000>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x00000] in <00000000000000000000000000000000>:0 
 
(Filename: currently not available on il2cpp Line: -1)
```

## Steps to reproduce

- New project
- Import `jillejr.newtonsoft.json-for-unity` via UPM
- Add following script to scene:

```csharp
void Start() {
    var json =
                "{\"token\":\"f61886be-6d2b-49ec-81da-1cb184729849\",\"expiresAt\":\"2020-10-29T17:58:36.408293Z\",\"createdAt\":\"2020-10-14T17:58:36.408293Z\"}";

    var accessTokenDto = JsonConvert.DeserializeObject<AccessTokenDto>(json);
    Debug.Log(accessTokenDto.Token);
}
```

- Build for WebGL
- Run the game

## Details

macOS Catalina 10.15.6

WebGL

12.0.301

2019.4.9f1

## Checklist

- [x] Shutdown Unity, deleted the /Library folder, opened project again in Unity, and problem still remains.
- [x] Checked to be using latest version of the package.

**Comment by JesseDunlap** [comment, marked as answer]
Hopefully I can answer my own question here. This turned out to be because Unity strips constructors and type parameters during the linking phase if they're not used. In order to negate this, you can use the [Preserve attribute](https://docs.unity3d.com/ScriptReference/Scripting.PreserveAttribute.html) or a [link.xml file](https://docs.unity3d.com/Manual/ManagedCodeStripping.html).

Adding `[Preserve]` attributes to my DTO class and its fields solved the issue.

```
[Preserve]
public class AccessTokenDto
{
    [Preserve]
    public string Token { get; set; }

    [Preserve]
    public DateTime ExpiresAt { get; set; }
    
    [Preserve]
    public DateTime CreatedAt { get; set; }
}
```

This may be something worthwhile adding to the README or Wiki. I'd be happy to do so if someone would like to point me to the right place to do it. Thanks!

**Comment by applejag** [comment]
Hi @JesseDunlap ! Thanks for reporting this :)

This is a known issue with the package, and I have added it to this repos wiki: <https://github.com/jilleJr/Newtonsoft.Json-for-Unity/wiki/Fix-AOT-compilation-errors>

I take full responsability in not making this knowledge more apparent. This is so common so I should indeed add this to the top of the README.md. It's good that you bring it up because I've forgotten to do so.

If you have any comments on the wiki pages, don't fray to create issues on the matter. The better, the better
